sdk/nodejs/output.ts

```diff
@@ -128,6 +130,20 @@ class OutputImpl<T> implements OutputInstance<T> {
             promise: Promise<T>,
             isKnown: Promise<boolean>,
             isSecret: Promise<boolean>) {
+
+        const value = Promise.all([isKnown, promise]).then(([known, val]) => (withUnknowns?: boolean) => {
+            // If the caller did not explicitly ask to see unknown values, simply return val.
+            if (!withUnknowns) {
+                return containsUnknowns(val) ? <T><any>undefined : val;
+            }
+
+            return val;
+        });
+
+        // Otherwise, we are only known if we are not explicitly unknown and the resolved value of the output
+        // contains no distinguished unknown values and if the value's promise is not rejected.
+        isKnown = Promise.all([isKnown, promise]).then(([known, val]) => known && !containsUnknowns(val));
```

Comment:
small favor.  one reason this is difficult to review is simply that there are so many similar names/parameters/locals in scope, and keeping track of each one is non trivial.  it would be great if we could extract out the two different Promise.all calls into helper functions.  It would make it very clear what was being overwritten here, and what the inner lambdas end up needing to capture.

Reply:
note: i'm fairly convinced this is all correct.  but this will simply make it clearer IMO.